### DISCUSSION TITLE
Compile sssp error

### DISCUSSION THREAD

**Discussion by FishieJ** [alibaba/GraphScope#1637] [Q&A] [answered]
I'm trying to run `graphscope.sssp` in Jupyter notebook and encounted the following error:
<img width="961" src="https://user-images.githubusercontent.com/18494026/170708865-1d914e29-98cb-4361-9b75-89cdae503dc6.png">
<img width="964" src="https://user-images.githubusercontent.com/18494026/170708898-1aca8545-ddaa-4996-a097-75c9f84d19cf.png">

CompilationError: Failed to compile app grape::SSSP<_GRAPH_TYPE> on platform system: Darwin
machine: x86_64
platform: macOS-11.5.2-x86_64-i386-64bit
uname: uname_result(system='Darwin', node='zhangsan.local', release='20.6.0', version='Darwin Kernel Version 20.6.0: Wed Jun 23 00:26:31 PDT 2021; root:xnu-7195.141.2\~5/RELEASE_X86_64', machine='x86_64')
kernel_ver: Darwin Kernel Version 20.6.0: Wed Jun 23 00:26:31 PDT 2021; root:xnu-7195.141.2\~5/RELEASE_X86_64
mac_ver: ('11.5.2', ('', '', ''), 'x86_64')
gcc_ver: b'gcc (Homebrew GCC 11.1.0) 11.1.0\nCopyright (C) 2021 Free Software Foundation, Inc.\nThis is free software; see the source for copying conditions.  There is NO\nwarranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n\n'
python_ver: 3.9.13

According to the above printed info (including mac_ver, gcc_ver, python_ver), I think I have satisfied all prequisites of installation. `graphscope.__version__` is '0.13.0'.

--------
Having tried gcc ver 7.3 in vain. It seems like graphscope still uses MacOS's default clang when compiling sssp.

**Comment by acezen** [comment]
Could you post your python code for reproducing the problem?

**Reply by FishieJ** [reply]
```python
import graphscope
sess = graphscope.session(cluster_type='hosts')
graph = sess.g()

import pandas as pd
df_v = pd.read_csv('/Downloaded_ldbc_sample/person_0_0.csv', sep='|')
df_v.to_csv('/PATH_TO/person_0_1.csv',sep=',',index=False)
df_e = pd.read_csv('/Downloaded_ldbc_sample/person_knows_person_0_0.csv', sep='|')
df_e.to_csv('/PATH_TO/person_knows_person_0_1.csv',sep=',',index=False)

graph = graph.add_vertices('/PATH_TO/person_0_1.csv', label='person')
graph = graph.add_edges('/PATH_TO/person_knows_person_0_1.csv',
                        src_label='person', dst_label='person')

import os, sys
sys.path.append('/PATH_TO/gcc-7.3/usr/local/bin')
os.system("gcc -v")
# Printed Info of gcc ver 7.3 here

result_sssp = graphscope.sssp(graph, src=6597069767117)
# Printed the error above. 
```
It seems like the printed path '/Library/Developer/CommandLineTools/SDKs/MacOSX11.3.sdk/usr/include/c++/v1/memory' belongs to MacOS's default clang. I'm not sure whether this is expected, and I haven't find a way to solve it yet.

**Comment by sighingnow** [comment, marked as answer]
SSSP cannot work on fragment with `empty type` as edata type. We need to add a check for such incompatibility.

@FishieJ For fragment with `empty type` as edata type, you could use the bfs algorithm.

**Reply by FishieJ** [reply]
Having no idea about what 'fragment' means and how `empty type` was created. Does that mean the input data, or the graph, has some specific feature?

**Reply by lidongze0629** [reply]
`Fragment` is a subgraph after the graph partition. The `SSSP` algorithm requires an attribute data of double type on the edge. 

```python3
>>> import graphscope
>>> from graphscope.dataset import load_p2p_network

>>> sess = graphscope.session(cluster_type="hosts", mode="eager")
>>> g = load_p2p_network(sess)   # load a property graph

>>> g.schema
type: VERTEX
Label: host
Properties: Property(0, weight, LONG), Property(1, id, LONG)

type: EDGE
Label: connect
Properties: Property(0, eid, LONG), Property(1, src_label_id, LONG), Property(2, dst_label_id, LONG), Property(3, dist, LONG)
Relations: [Relation(source='host', destination='host')]

>>> # project the property graph to a simple graph, sssp need `dist` property(double type) on the edge  
>>> pg = g.project(vertices={"host": ["id"]}, edges={"connect": ["dist"]})

>>> c = graphscope.sssp(pg, src=6)
```

**Reply by FishieJ** [reply]
> 

This is helpful. Thanks.
